src/api/forms/repositories/aggregation/form-metadata-aggregation.js

```diff
@@ -0,0 +1,225 @@
+              {
+                case: {
+                  $regexMatch: {
+                    input: '$title',
+                    regex: `\\b${escapeRegExp(searchTerm)}\\b`,
+                    options: 'i'
+                  }
+                },
+                then: 2
+              },
+              // Rank 3: Partial substring match in the title (case-insensitive)
+              {
+                case: {
+                  $regexMatch: {
+                    input: '$title',
+                    regex: escapeRegExp(searchTerm),
+                    options: 'i'
+                  }
+                },
+                then: 3
+              }
+            ],
+            default: 4
+          }
+        })
+      }
+    })
+  } else {
+    pipeline.push({
+      $addFields: {
+        matchScore: 0
+      }
+    })
+  }
+}
+
+/**
+ * Adds the date field stage to the pipeline to extract date only from 'updatedAt'.
+ * @param {PipelineStage[]} pipeline - The aggregation pipeline stages.
+ */
+export function addDateFieldStage(pipeline) {
+  const dateFieldStage = {
+    $addFields: {
+      updatedDateOnly: {
+        $dateToString: {
+          format: '%Y-%m-%d',
+          date: '$updatedAt',
+          timezone: 'UTC'
+        }
+      }
+    }
+  }
+  pipeline.push(dateFieldStage)
+}
+
+/**
+ * Adds the sorting stage to the pipeline based on the sortBy parameter.
+ * @param {PipelineStage[]} pipeline - The aggregation pipeline stages.
+ * @param {string} sortBy - Field to sort by ('updatedAt' or 'title').
+ * @param {string} order - Sort order ('asc' or 'desc').
+ * @returns {CollationOptions | null} The collation options if necessary.
+ */
+export function addSortingStage(pipeline, sortBy, order) {
+  const sortOrder = order === 'asc' ? 1 : -1
+  const collation = { locale: 'en', strength: 1 }
+
+  switch (sortBy) {
+    case 'title':
+      pipeline.push({
+        $sort: {
+          // Primary sort is title
+          title: sortOrder,
+          // Then newest first if titles tie
+          updatedDateOnly: -1,
+          // Then alphabetical (case-insensitive) on displayName
+          'updatedBy.displayName': 1
+        }
+      })
+      break
+
+    case 'updatedAt':
+      pipeline.push({
+        $sort: {
+          updatedDateOnly: sortOrder,
+          'updatedBy.displayName': 1
+        }
+      })
+      break
+
+    default:
+      pipeline.push({
+        $sort: {
+          updatedDateOnly: -1,
+          'updatedBy.displayName': 1
+        }
+      })
+      break
+  }
+
+  return collation
+}
+
+/**
+ * @typedef {object} FilterConditions
+ * @property {{ $regex: RegExp }} [title] - Optional MongoDB regex query for title matching
+ */
+
+/**
+ * @typedef {{
+ *   case: {
+ *     $eq?: [{$toLower: string} | string, string],
+ *     $regexMatch?: {
+ *       input: string,
+ *       regex: string,
+ *       options: string
+ *     }
+ *   },
+ *   then: number
+ * }} SwitchBranch
+ */
```

Comment:
Thanks for adding these types @mokhld

Would you mind moving them into a **types.js** file like the others?

Great too if you can use the simpler non-nested format with descriptions:

```suggestion
/**
 * @typedef {object} SwitchCase
 * @property {[{ $toLower: string } | string, string]} [$eq] - Description here
 * @property { input: string, regex: string, options: string } [$regexMatch] - Description here
 */

/**
 * @typedef {object} SwitchBranch
 * @property {SwitchCase} case - Description here
 * @property {number} then - Description here
 */
```

Reply:
That's been moved and simplified. It was quite verbose before, mainly for my benefit as constructing these were a pain 😩